### PR TITLE
Fix incorrectly wrapped arrow functions with return types

### LINKED ISSUE #10923: Formatting for a typescript function call where the last element is a function with a return type annotation that is a generic type has inconsistent behavior.
**Prettier 2.3.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAbaBzAZgVymGAS2gHVCYALAYQENVUAhGsAawAoAdEAI1Rt5pcANAAIsECNxoAnKQC8BivqLaVCUDEhEBnGNPUYAlFsgBbAA6o4ADwA8MAJ7m4t7RFNw1GgHzeRAXj9gDigREUgAEzg2QwBuEIBfOJAhEAhzImhtZFAZaQgAdwAFGQRslDoCmgds1O5pZhZPAGVzZgNkPRw4VLhTbjgIqIiAGRoNHBoMOAAxCGlTGhgiDWQQGhwYCBSQChhTVBIKcjhtNrA4ZrLyQgA3cgc1sG1akHVtOGkYIoaMReQsHQPqkAFbaawMBqsFo0DwjdRwAFAnogMHWZoGKwARRwEHgSNQwJAbWkH2ka0czm0YH0GR25n0sDIEUoyAAHAAGVIMiAfEgNcxrBmnT63RGpACOuPgP3S5XW2gAtFA4INBjtpHApYRNT8pv8kIDCSiPqZCJ1pN1UtpMXAcXjEYbkakYPxmaykAAmF0NQioAxUdwGkCnACsOxwHwAKvxykaibdugBJKBRWDNGmEDIAQVTzUcVgJHwSCSAA)

```sh
--parser typescript
```

**Input:**

```tsx
longfunctionWithCallBack("blabla", foobarbazblablabla, (thing: string): complex<type<something>> => {
  code();
});
```

**Output:**

```tsx
longfunctionWithCallBack("blabla", foobarbazblablabla, (thing: string): complex<
  type<something>
> => {
  code();
});

```

**Output in 2.2.1:**

```typescript
longfunctionWithCallBack(
  "blabla",
  foobarbazblablabla,
  (thing: string): complex<type<something>> => {
    code();
  }
);

```

**Without nested generic type annotation**:

**Input**

```typescript
longfunctionWithCallBack("blabla", foobarbazblablabla, (thing: string): complex<type> => {
  code();
});
```

**output**

```typescript
longfunctionWithCallBack(
  "blabla",
  foobarbazblablabla,
  (thing: string): complex<type> => {
    code();
  }
);

```

**Expected behavior:**
I personally find the old behavior/behavior with no nested generics more readable.

### PR DESCRIPTION
**Pull request by thorn0**
## Description

fixes #10923

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/print/function.js [hunk 1]
```diff
@@ -176,16 +178,24 @@ function printArrowFunctionSignature(path, options, print, args) {
   if (shouldPrintParamsWithoutParens(path, options)) {
     parts.push(print(["params", 0]));
   } else {
+    const expandArg = args && (args.expandLastArg || args.expandFirstArg);
+    let returnTypeDoc = printReturnType(path, print, options);
+    if (expandArg) {
+      if (willBreak(returnTypeDoc)) {
+        throw new ArgExpansionBailout();
+      }
```
Comment (sosukesuzuki) [inline src/language-js/print/function.js:186]: Can you add tests that matches this condition?

Comment (thorn0) [reply src/language-js/print/function.js:186]: done

File: tests/format/typescript/argument-expansion/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -85,3 +85,75 @@ const bar8 = [1, 2, 3].reduce(
...
+    code();
+  }
+);
+
+longfunctionWithCall1(
+  "bla",
+  foo,
+  (
+    thing: string
+  ): complex<
```
Comment (sosukesuzuki) [inline tests/format/typescript/argument-expansion/__snapshots__/jsfmt.spec.js.snap:150]: It may be a trivial matter, but this line breaks is expected?

Comment (thorn0) [reply tests/format/typescript/argument-expansion/__snapshots__/jsfmt.spec.js.snap:150]: It's not, but it's a separate issue

Comment (sosukesuzuki) [reply tests/format/typescript/argument-expansion/__snapshots__/jsfmt.spec.js.snap:150]: ok

**Review by sosukesuzuki — APPROVED** [review]
I've tried the same issue, but I have no good solution. Thank you!

**Comment by thorn0** [comment]
I still remembered how these things work after working on #10712. BTW, first I thought this is a regression in 2.3, but it turned out it's an old bug.

**Pull request merged**